library/rsa.c

```diff
@@ -1429,15 +1442,27 @@ int mbedtls_rsa_rsassa_pkcs1_v15_verify( mbedtls_rsa_context *ctx,
     if( asn1_len != hashlen )
         return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
 
-    if( memcmp( p, hash, hashlen ) != 0 )
-        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
+    ret = MBEDTLS_ERR_RSA_VERIFY_FAILED;
+    ret = memcmp( p, hash, hashlen );
 
-    p += hashlen;
+    if( ret == 0 )
+    {
+        if( ret != 0)
+            return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
 
-    if( p != end )
-        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
+        p += hashlen;
+        vp = p;
 
-    return( 0 );
+        if( vp == end )
```

Comment:
This check (error out if `p != end`) already ensures that the hash comes at the end of the signature buffer. Isn't this enough to rule out the Bleichenbacher attack (assuming no perturbation of code execution)? The code did already check that the buffer consists of header, padding, ASN.1 tags, hash value and nothing else.